urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/common/number/CohoAPR.java

```diff
@@ -0,0 +1,528 @@
+	 * ceil(log2(x)) = (x-1).bitLength;       
+	 */
+	@Override
+	public double doubleValue(ROUNDMODE mode){
+		int cmp = numerator.compareTo(BigInteger.ZERO);
+		if(cmp==0){
+			return 0;
+		}
+		if(cmp<0){//negative
+			ROUNDMODE negMode ;
+			switch(mode){
+			case CEIL:
+				negMode = ROUNDMODE.FLOOR;
+				break;
+			case FLOOR:
+				negMode = ROUNDMODE.CEIL;
+				break;
+			case ZERO:
+				negMode = ROUNDMODE.ZERO;
+				break;
+			default:
+				negMode = ROUNDMODE.NEAR;
+			}
+			return -negate().doubleValue(negMode);
+		}
+		
+		cmp = numerator.compareTo(denominator);
+		if(cmp ==0){
+			return 1;
+		}
+		
+		//estimate the exp
+		int exp;
+		if(cmp>0){//n>d normalized number
+			//estimate exp;
+			BigInteger div = floor(numerator,denominator);//estimate exp by floor.
+			exp = floorLog2(div);//floor of log2(div)
+		}else{
+			BigInteger div = ceil(denominator, numerator);
+			exp = ceilLog2(div);
+			exp = -exp;
+		}
+
+		BigInteger M = null;
+		boolean denomalized = false;
+		for(boolean round=true; round; ){
+			if(exp<-1022){//denomalized number 0.f*2^-1022
+				denomalized = true;
+				exp = -1022;
+			}
+			switch(mode){
+			case CEIL:
+				if(exp>0){
+					M = ceil(leftShift(numerator,52),leftShift(denominator,exp));
+				}else{
+					M = ceil(leftShift(numerator,52-exp),denominator);	
+				}
+				break;
+			case FLOOR:
+			case ZERO:
+				if(exp>0){
+					M = floor(leftShift(numerator,52),leftShift(denominator,exp));
+				}else{
+					M = floor(leftShift(numerator,52-exp),denominator);				
+				}
+				break;
+			default:
+				if(exp>0){
+					M = round(leftShift(numerator,52),leftShift(denominator,exp));
+				}else{
+					M = round(leftShift(numerator,52-exp),denominator);
+				}
+			}
+			if(exp>=1024){
+				//or return infinity?
+				throw new ArithmeticException("Out of range of double value"+exp+"\t"+numerator.divide(denominator).doubleValue());
+			}
+			if(M.bitLength()==54){
+				exp = exp+1;//reround use the new exp
+			}else{
+				round = false;
+				if(M.bitLength()>54)
+					throw new RuntimeException("CohoAPR.round(): Algorithm error "+M.bitLength());
+				if(M.bitLength()!=53 && !denomalized)
+					throw new RuntimeException("CohoAPR.round(): The significant should be 53 bits "+M.bitLength());
+			}
+		}
+
+		//use m and exp to construct the double
+		long m = M.longValue();
+		long l= ( (m&0x000fffffffffffffL) | ((long)(exp+BIAS))<<52 );
+		if(denomalized)
+			l = (m&0x000fffffffffffffL);//exp should be zero not 1
+		return Double.longBitsToDouble(l);
+	}
+	public static boolean test(){
+		boolean result = true;
+		int n = 100;
+		for(int i=0; i<n; i++){//test n>d
+			CohoDouble d = CohoDouble.zero.random();
+			double dd = d.doubleValue()*Double.MAX_VALUE;
+			CohoAPR apr = CohoAPR.create(d).mult(CohoAPR.create(Double.MAX_VALUE));
+			if(dd!=apr.doubleValue(ROUNDMODE.NEAR)){
+				System.out.println("original "+dd+" vs. "+apr.doubleValue(ROUNDMODE.NEAR));
+				result = false;
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 427:435, 445:455.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.